File: lever/src/main/java/com/pin/lever/DeferredRequestCreator.java
package com.pin.lever;

import android.support.annotation.VisibleForTesting;
import android.view.View;
import android.view.View.OnAttachStateChangeListener;
import android.view.ViewTreeObserver;
import android.view.ViewTreeObserver.OnPreDrawListener;
import android.widget.ImageView;
import java.lang.ref.WeakReference;

class DeferredRequestCreator implements OnPreDrawListener, OnAttachStateChangeListener {
    private final RequestCreator creator;
    @VisibleForTesting
    final WeakReference<ImageView> target;
    @VisibleForTesting Callback callback;

    DeferredRequestCreator(RequestCreator creator, ImageView target, Callback callback) {
        this.creator = creator;
        this.target = new WeakReference<>(target);
        this.callback = callback;

        target.addOnAttachStateChangeListener(this);

        // Only add the pre-draw listener if the view is already attached
        if (target.getWindowToken() != null) {
            onViewAttachedToWindow(target);
        }
    }

    @Override public void onViewAttachedToWindow(View view) {
        view.getViewTreeObserver().addOnPreDrawListener(this);
    }

    @Override public void onViewDetachedFromWindow(View view) {
        view.getViewTreeObserver().removeOnPreDrawListener(this);
    }

    @Override public boolean onPreDraw() {
        ImageView target = this.target.get();
        if (target == null) {
            return true;
        }

        ViewTreeObserver vto = target.getViewTreeObserver();
        if (!vto.isAlive()) {
            return true;
        }

        int width = target.getWidth();
        int height = target.getHeight();

        if (width <= 0 || height <= 0) {
            return true;
        }

        target.removeOnAttachStateChangeListener(this);
        vto.removeOnPreDrawListener(this);
        this.target.clear();

        this.creator.unfit().resize(width, height).into(target, callback);
        return true;
    }

    void cancel() {
        creator.clearTag();
        callback = null;

        ImageView target = this.target.get();
        if (target == null) {
            return;
        }
        this.target.clear();

        target.removeOnAttachStateChangeListener(this);

        ViewTreeObserver vto = target.getViewTreeObserver();
        if (vto.isAlive()) {
            vto.removeOnPreDrawListener(this);
        }
    }

    Object getTag() {
        return creator.getTag();
    }
}

File: lever/src/main/java/com/pin/lever/DeferredRequestCreator.kt
package com.pin.lever

import android.support.annotation.VisibleForTesting
import android.view.View
import android.view.ViewTreeObserver
import android.widget.ImageView
import java.lang.ref.WeakReference

class DeferredRequestCreator(
    private val creator: RequestCreator,
    target: ImageView,
    callback: Callback?
) : ViewTreeObserver.OnPreDrawListener, View.OnAttachStateChangeListener {
    @VisibleForTesting
    val target: WeakReference<ImageView>
    @VisibleForTesting
    var callback: Callback?

    override fun onViewAttachedToWindow(view: View) {
        view.viewTreeObserver.addOnPreDrawListener(this)
    }

    override fun onViewDetachedFromWindow(view: View) {
        view.viewTreeObserver.removeOnPreDrawListener(this)
    }

    override fun onPreDraw(): Boolean {
        val target = target.get() ?: return true
        val vto = target.viewTreeObserver
        if (!vto.isAlive) {
            return true
        }
        val width = target.width
        val height = target.height
        if (width <= 0 || height <= 0) {
            return true
        }
        target.removeOnAttachStateChangeListener(this)
        vto.removeOnPreDrawListener(this)
        this.target.clear()
        creator.unfit().resize(width, height).into(target, callback)
        return true
    }

    fun cancel() {
        creator.clearTag()
        callback = null
        val target = target.get() ?: return
        this.target.clear()
        target.removeOnAttachStateChangeListener(this)
        val vto = target.viewTreeObserver
        if (vto.isAlive) {
            vto.removeOnPreDrawListener(this)
        }
    }

    val tag: Any?
        get() = creator.tag

    init {
        this.target = WeakReference(target)
        this.callback = callback
        target.addOnAttachStateChangeListener(this)
        // Only add the pre-draw listener if the view is already attached
        if (target.windowToken != null) {
            onViewAttachedToWindow(target)
        }
    }
}